File: workflow/scripts/selection_methods/gene_selection_geneBasis.R
# wrapper for gene selection with geneBasis

# load libraries
#suppressPackageStartupMessages(library(devtools))
#install_github('MarioniLab/geneBasisR')
suppressPackageStartupMessages(library(geneBasisR))
suppressPackageStartupMessages(library(zellkonverter))
suppressPackageStartupMessages(library(SingleCellExperiment))

# parse arguments

args <- commandArgs(trailingOnly = TRUE)
n <- as.integer(args[1])
prep_adata_path <-  args[2]
output_path <- args[3]

if(args[4] == 'None'){
  genes_base <- NULL
}else{
  genes_base <- strsplit(args[4], split = " ")
}
batch <- args[5]
if(batch == "None"){
  batch <- NULL
}
neigh <- as.integer(args[6])
if(is.na(neigh)){
  neigh <- 5
}
mink <- as.integer(args[7])
if(mink == "None"){
  mink <- 3
}
npc_selection <- as.integer(args[8])
if(is.na(npc_selection)){
  npc_selection <- NULL
}
npc_all <- as.integer(args[9])
if(is.na(npc_all)){
  npc_all <- 50
}
if(args[10] == 'None'){
  genes_discard <- NULL
}else{
  genes_discard <- strsplit(args[10], split = " ")
}
if(args[11] == 'None'){
  genes_discard_prefix <- NULL
}else{
  genes_discard_prefix <- strsplit(args[11], split = " ")
}
verbose <- as.logical(args[12])
if(verbose == "None"){
  verbose <- TRUE
}
#conda_env <- args[13]

stat_all <- NULL
#' @param stat_all If True graph and corresponding Minkowski distances have been calculated prior to search, provide this data here.
#' It can be useful if gene_search is desired to be recycled (e.g. for selecting multiple libraries with different inputs such as n_genes_total and genes_base)
#' Ensure that colnames = c("gene", "dist_all"). Default stat_all=NULL - in case this info is not supplied.


# minimal example:
# conda_env <- "/big/st/strasserl/gene_selection_methods/parent_dir/environments/venv_genebasis"
# prep_adata_path <- "/big/st/strasserl/gene_selection_methods/parent_dir/selections/tmp/preprocessed_adata.h5ad"
# output_path <- "/big/st/strasserl/gene_selection_methods/parent_dir/selections/tmp/genebasis_tmp.tsv"

## specify conda env
suppressPackageStartupMessages(library(reticulate))
#use_condaenv(condaenv = conda_env, required = T)
use_python("/usr/bin/python")

# read input (log normalized HVGs)
# creates a new conda env in ~/.cache/basilisk/1.2.1/zellkonverter-1.0.3/anndata_env
input <- readH5AD(prep_adata_path)   # basilisk
# rows: genes, columns: cells, $cell: unqique cell ids, assay called logcounts
logcounts(input) <- assay(input)
input$cell <- colnames(input)

#' n_row = 1000
#' n_col = 100
#' sce = SingleCellExperiment(assays = list(logcounts = matrix(rnorm(n_row*n_col), ncol=n_col)))
#' rownames(sce) = as.factor(1:n_row)
#' colnames(sce) = c(1:n_col)
#' sce$cell = colnames(sce)
#' genes = rownames(sce)
#' out = gene_search(sce, n_genes_total = 5)
#'
#' SingleCellExperiment object, wehre normalized counts stored in 'logcounts' assay

selection <- gene_search(input,
                        n_genes_total = n,
                        genes_base = genes_base,
                        batch = batch,
                        n.neigh = neigh,
                        p.minkowski = mink,
                        nPC.selection = npc_selection,
                        nPC.all = npc_all,
                        genes.discard = genes_discard,
                        genes.discard_prefix = genes_discard_prefix,
                        verbose = verbose,
                        stat_all = stat_all)

print(selection)

# write (tmp) output file
write.table(selection, file=output_path, quote=F, sep="\t", row.names=F)

File: workflow/scripts/selection_methods/gene_selection_genebasis.py
import pandas as pd
import numpy as np
import scanpy as sc
import time
import subprocess
import os

from selection_methods.gene_selection_shared import *


def select_genes_genebasis(n, adata, tmp_dir, #conda_env, 
                           r_exe="Rscript", #size_factors="size_factors",
                           genes_base=None, batch=None, neigh=5, mink=3, npc_selection=None, npc_all=50,
                           genes_discard=None, genes_discard_prefix=None, verbose=True):
    """

    Parameters
    ----------
    n: int
        Number of genes to select
    adata: Anndata
        Anndata object with raw counts in adata.X and size factors in adata.obs[size_factors]
    tmp_dir: String
        path where tmp files should be stored
    conda_env:
        name (or prefix) of conda environment
    r_exe:
        path to Rscript (in conda env)
    size_factors: String (default: "size_factors")
        name of adata.obs were size factors are stored
    genes_base: list of Strings (default: None)
        Character vector specifying base genes to construct first Selection graph. Default=NULL in case no genes are supplied.
    batch: String (default: None)
        Name of the field in colData(sce) to specify batch.
    neigh: int (default: 5)
        Positive integer > 1, specifying number of neighbors to use for kNN-graph.
    mink: int (default: 3)
        p.minkowski Order of Minkowski distance.
    npc_selection: int (default: None)
        Scalar specifying number of PCs to use for Selection Graphs.
    npc_all: int (default: 50)
        Scalar specifying number of PCs to use for True Graph.
    genes_discard: list of Strings (default: None)
        Character vector containing genes to be excluded from candidates (note that they still will be used for graphs construction. If you want to exclude them from graph construction as well, just discard them prior in sce object). Default = NULL and no genes will be discarded.
    genes_discard_prefix: list of Strings (default: None)
        Character vector containing prefixes of genes to be excluded (e.g. Rpl for L ribosomal proteins. Note that they still will be used for graphs construction. If you want to exclude them from graph construction as well, just discard them prior in sce object). Default = NULL and no genes will be discarded.
    verbose: Boolean (default: True)
        Boolean identifying whether intermediate print outputs should be returned
    """

    #adata = preprocess_adata(adata, size_factors=size_factors)
    preprocessed_adata_path = os.path.join(tmp_dir, "preprocessed_adata.h5ad")
    adata.write_h5ad(preprocessed_adata_path)
    del adata  # need to delete the anndata because otherwise the h5ad file is locked and can't be read by the R script

    select_scripts_dir = os.path.dirname(os.path.abspath(__file__))

    start = time.time()
    command = r_exe + f" {select_scripts_dir}/gene_selection_geneBasis.R " + " ".join([str(n),  # 1
                                                                                   preprocessed_adata_path,  # 2
                                                                                   os.path.join(tmp_dir, "genebasis_tmp.tsv"),  # 3
                                                                                   str(" ".join(genes_base)) if genes_base is not None else str(None),  # 4
                                                                                   str(batch),  # 5
                                                                                   str(neigh),  # 6
                                                                                   str(mink),  # 7
                                                                                   str(npc_selection),  # 8
                                                                                   str(npc_all),  # 9
                                                                                   str(" ".join(genes_discard)) if genes_discard is not None else str(None),  # 10
                                                                                   str(" ".join(genes_discard_prefix)) if genes_discard_prefix is not None else str(None),  # 11
                                                                                   str(verbose)#,  # 12
                                                                                   #str(conda_env)  # 13
                                                                                   ])

    # example command:
    # ./environments/venv_genebasis/bin/Rscript selection_methods/gene_selection_geneBasis.R 25 selections/tmp/preprocessed_adata.h5ad selections/tmp None None 5 3 None 50 None None True

    process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE)
    process.wait()
    end = time.time()
    took = end - start
    adata = sc.read(preprocessed_adata_path)
    selectedGenes = list(pd.read_csv(os.path.join(tmp_dir,"genebasis_tmp.tsv"), sep="\t")['gene'])
    adata.var['idx'] = range(adata.shape[1])
    idx = [int(adata.var['idx'][adata.var.index == x]) for x in selectedGenes]
    selection = pd.DataFrame({"idx": idx, "selection": selectedGenes})


    return selection, took